### DISCUSSION TITLE
record is missing the primary key but defined in the model ?

### DISCUSSION THREAD

**Discussion by SPG87** [CodeDredd/pinia-orm#1717] [Q&A] [answered]
Hey everyone, I'm a beginner here working with Vue and Pinia-ORM, and I have the following problem that I can't solve on my own.

I currently have two models in an app: clients and projects. I call them using a fetch request and receive the response with no problem. I can also save them in their repos and everything fires perfectly, except for when I try to define the relationship between the two. When i do i get the following error message:

`Uncaught (in promise) Error: [Pinia ORM] The record is missing the primary key. If you want to persist record without the primary key, please define the primary key field with the uid attribute.`

I adapted them from SQLAlchemy models, here they are:
Project Model:

    id = mapped_column(
        sa_types.Integer, primary_key=True, autoincrement="auto")
    client_id = mapped_column(ForeignKey('client.id'), nullable=False)
    client = relationship("ClientModel")
    name = mapped_column(sa_types.Text, nullable=False)
    description = mapped_column(sa_types.Text, nullable=True)
    active = mapped_column(
        sa_types.Boolean, nullable=False, default=True, server_default="true")
    start_date = mapped_column(sa_types.Date, nullable=True)
    end_date = mapped_column(sa_types.Date, nullable=True)
    status_id = mapped_column(
        sa_types.Integer, nullable=False, default=1, server_default="1")
    type_id = mapped_column(
        sa_types.Integer, nullable=False, default=2, server_default="2")
    budget = mapped_column(
        sa_types.DECIMAL, nullable=False, default=0, server_default="0")

    def __repr__(self):
        return (f"{self.id} {self.client_id} {self.name} {self.description} "
                f"{self.active} {self.start_date} {self.end_date} "
                f"{self.status_id} {self.type_id} {self.budget}")

Client Model:

    id = mapped_column(
        sa_types.Integer, primary_key=True, autoincrement="auto")
    name = mapped_column(sa_types.Text, nullable=False)
    active = mapped_column(sa_types.Boolean, nullable=False)
    projects: Mapped[List[ProjectModel]] = relationship(
        back_populates="client")

    def __repr__(self):
        return f"{self.id} {self.name} {self.active}"

Pinia-ORM Models
Projects:

    import { Model } from "pinia-orm";
    import { Attr, BelongsTo, Bool, Num, Str, Uid } from 'pinia-orm/dist/decorators'
    import clients from '@app/stores/pinia-orm/models/ClientModel'

    export default class ProjectModel extends Model {
    static entity = 'projects'
    static primaryKey = 'project_id'

    @Uid() declare project_id: number
    @BelongsTo(() => clients, 'client_id') declare client_id: clients
    @Str(null, { notNullable: true }) declare name: string
    @Str(null, { notNullable: true }) declare description: string | null
    @Bool(true) declare active: boolean
    @Attr() declare start_date: Date | null
    @Attr() declare end_date: Date | null
    @Num(ProjectModelHelpers.serverDefaultValues.statusIdDefaultValue, { notNullable: true })
    declare status_id: number

    @Num(ProjectModelHelpers.serverDefaultValues.typeIdDefaultValue, { notNullable: true })
    declare type_id: number

    @Num(ProjectModelHelpers.serverDefaultValues.budgetDefaultValue, { notNullable: true })
    declare budget: number
    }

Clients:

    import { Model } from "pinia-orm";
    import { Bool, HasMany, Str, Uid } from 'pinia-orm/dist/decorators'
    import projects from '@app/stores/pinia-orm/models/ProjectModel'

    export default class ClientModel extends Model {
    static entity = 'clients'
    static primaryKey = 'client_id'

    @Uid() declare client_id: number
    @Str(null, { notNullable: true }) declare name: string
    @Bool(null, { notNullable: true }) declare active: boolean
    @HasMany(() => projects, 'client_id') declare projects: projects[]
     }

What am I missing here?

**Comment by ldiebold** [comment, marked as answer]
At a glance, it looks like you may need a `client_id` on the "Projects" model. I'm going to rewrite some of this code, and change the naming to make it more inline with what you'd usually see with PiniaORM. Specifically, changing this line:
```ts
@BelongsTo(() => clients, 'client_id') declare client_id: clients
```

to this:
```ts
@BelongsTo(() => Client, 'client_id') declare client: client
```

---

`ProjectModel.ts` (note, this is usually pascal, singular. You may also want to drop the word `Model` so its reads `Project.ts`)
```ts
import { Model } from "pinia-orm";
import { Attr, BelongsTo, Bool, Num, Str, Uid } from 'pinia-orm/dist/decorators'
import Client from '@app/stores/pinia-orm/models/ClientModel'

export default class ProjectModel extends Model {
  static entity = 'projects'
  static primaryKey = 'project_id'
  
  @Uid() declare project_id: number
  @BelongsTo(() => Client, 'client_id') declare client: Client | null
  // the below line is new. We must also declare the client_id.
  @Attr() declare client_id: string
  @Str(null, { notNullable: true }) declare name: string
  @Str(null, { notNullable: true }) declare description: string | null
  @Bool(true) declare active: boolean
  @Attr() declare start_date: Date | null
  @Attr() declare end_date: Date | null
  @Num(ProjectModelHelpers.serverDefaultValues.statusIdDefaultValue, { notNullable: true })
  declare status_id: number
  
  @Num(ProjectModelHelpers.serverDefaultValues.typeIdDefaultValue, { notNullable: true })
  declare type_id: number
  
  @Num(ProjectModelHelpers.serverDefaultValues.budgetDefaultValue, { notNullable: true })
  declare budget: number
}
```

`ClientModel.ts`
```ts
import { Model } from "pinia-orm";
import { Bool, HasMany, Str, Uid } from 'pinia-orm/dist/decorators'
import Project from '@app/stores/pinia-orm/models/ProjectModel'

export default class ClientModel extends Model {
  static entity = 'clients'
  static primaryKey = 'client_id'
  
  @Uid() declare client_id: number
  @Str(null, { notNullable: true }) declare name: string
  @Bool(null, { notNullable: true }) declare active: boolean
  @HasMany(() => projects, 'client_id') declare projects: Project[]
 }
```

Also hot tip! To add syntax highlighting to your code in github, follow your first three bacticks with the language.
So in this case that would be: ```ts

**Reply by SPG87** [reply]
Thanks a lot man !!
